psi4/src/psi4/libscf_solver/hf.cc

```diff
@@ -424,55 +424,71 @@ void HF::find_occupation() {
         MOM();
     } else {
-        // We first find the aufbau occupation.
-        // We then take our orbitals as the aufbau orbitals within the occupation.
-        std::vector<std::pair<double, int> > pairs_a;
-        std::vector<std::pair<double, int> > pairs_b;
-        for (int h = 0; h < epsilon_a_->nirrep(); ++h) {
-            for (int i = 0; i < epsilon_a_->dimpi()[h]; ++i) {
-                pairs_a.push_back(std::make_pair(epsilon_a_->get(h, i), h));
-            }
-        }
-        sort(pairs_a.begin(), pairs_a.end());
-
-        // Do we need to sort beta?
-        if (multiplicity_ == 1) {
-            pairs_b = pairs_a;
+        if (!input_docc_ && !input_socc_) {
+            assert(nirrep_ == epsilon_a_->nirrep());
+            assert(nirrep_ == epsilon_b_->nirrep());
 
-        } else {
-            for (int h = 0; h < epsilon_b_->nirrep(); ++h) {
+            // The occupations are determined by the Aufbau
+            // principle. We first collect all the orbital energies and
+            // sort them in increasing order
+            std::vector<std::pair<double, int> > pairs_a;
+            for (int h = 0; h < nirrep_; ++h) {
+                for (int i = 0; i < epsilon_a_->dimpi()[h]; ++i) {
+                    pairs_a.push_back(std::make_pair(epsilon_a_->get(h, i), h));
+                }
+            }
+            sort(pairs_a.begin(), pairs_a.end());
+            // Same for beta electrons
+            std::vector<std::pair<double, int> > pairs_b;
+            for (int h = 0; h < nirrep_; ++h) {
                 for (int i = 0; i < epsilon_b_->dimpi()[h]; ++i) {
                     pairs_b.push_back(std::make_pair(epsilon_b_->get(h, i), h));
                 }
             }
             sort(pairs_b.begin(), pairs_b.end());
-        }
 
-        if (!input_docc_ && !input_socc_) {
-            // Sanity check
+            // Sanity check: we must have at least one orbital per electron
             if ((size_t)std::max(nalpha_, nbeta_) > pairs_a.size())
                 throw PSIEXCEPTION("Not enough basis functions to satisfy requested occupancies");
 
-            // Alpha
-            memset(nalphapi_, 0, sizeof(int) * epsilon_a_->nirrep());
+            // Reset occupations
+            for (int h = 0; h < nirrep_; ++h) {
+                nalphapi_[h] = 0;
+                nbetapi_[h] = 0;
+            }
+            // Occupy the lowest nalpha orbitals
             for (int i = 0; i < nalpha_; ++i) nalphapi_[pairs_a[i].second]++;
-
-            // Beta
-            memset(nbetapi_, 0, sizeof(int) * epsilon_b_->nirrep());
+            // Occupy the lowest nbeta electrons
             for (int i = 0; i < nbeta_; ++i) nbetapi_[pairs_b[i].second]++;
         }
 
-        int old_socc[8];
-        int old_docc[8];
+        int old_socc[nirrep_];
+        int old_docc[nirrep_];
         for (int h = 0; h < nirrep_; ++h) {
             old_socc[h] = soccpi_[h];
             old_docc[h] = doccpi_[h];
         }
```

Comment:
The reason the old code had `8` instead of `nirrep_` was because the old way of doing it created a statically allocated array. But`nirrep_` isn't known until runtime, so I'm worried that _even without new_, we're dynamically allocating.

`soccpi_` and `doccpi_` are Dimension objects, so why not have `old_socc` and `old_docc` by Dimension objects created by an implicit copy constructor? No chance of memory leaks and less code.

Reply:
Sure. Using `Dimension` also resulted in other negative LOC.